core-api/src/main/java/com/optimizely/ab/config/audience/MatchType.java

```diff
@@ -0,0 +1,187 @@
+package com.optimizely.ab.config.audience;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+interface LeafMatcher {
+    public Boolean eval(Object otherValue);
+}
+
+abstract class LeafMatch<T> implements LeafMatcher {
+    T convert(Object o) {
+        try {
+            T rv = (T)o;
+            return rv;
+        } catch(java.lang.ClassCastException e) {
+            return null;
+        }
+    }
+}
+
+class NullMatch extends LeafMatch<Object> {
+    @SuppressFBWarnings("URF_UNREAD_FIELD")
+    Object value;
+    protected NullMatch() {
+        this.value = null;
+    }
+
+    public @Nullable Boolean eval(Object otherValue) {
+        return null;
+    }
+}
+
+class ExactMatch<T> extends LeafMatch<T> {
+    T value;
+    protected ExactMatch(T value) {
+        this.value = value;
+    }
+
+    public @Nullable Boolean eval(Object otherValue) {
+        if (!value.getClass().isInstance(otherValue)) return null;
+        return value.equals(convert(otherValue));
+    }
+}
+
+/**
+ * This is a temporary class.  It mimics the current behaviour for
+ * custom dimension.  This will be dropped for ExactMatch and the unit tests need to be fixed.
+ * @param <T>
+ */
+class CustomDimensionMatch<T> extends LeafMatch<T> {
```

Comment:
We don't support custom dimensions anymore, even in 1.x. I think we can safely omit this.